File: ClassesDemo/src/Day.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;

public class Day
{
    private LocalDate date;
    private ArrayList<Task> optionalTasks;
    private ArrayList<Period> schedule;
    private double availableHours;
    private double totalValue;
    
    public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy");

    public Day(LocalDate date)
    {
        this.date = date;
        this.optionalTasks = new ArrayList<>();
        this.schedule = new ArrayList<>();
        this.availableHours = 7;
        this.totalValue = 0;
    }

    // ===========================================================================

    public LocalDate getDate()
    {
        return date;
    }

    public ArrayList<Task> getOptionalTasks()
    {
        return optionalTasks;
    }

    public ArrayList<Period> getSchedule()
    {
        return schedule;
    }

    public double getAvailableHours()
    {
        return availableHours;
    }

    public double getTotalValue()
    {
        return totalValue;
    }

    // ===========================================================================

    public void setDate(LocalDate date)
    {
        this.date = date;
    }

    public void setOptionalTasks(ArrayList<Task> optionalTasks)
    {
        this.optionalTasks = optionalTasks;
    }

    public void setSchedule(ArrayList<Period> schedule)
    {
        this.schedule = schedule;
    }

    public void setAvailableHours(double availableHours)
    {
        this.availableHours = availableHours;
    }

    public void setTotalValue(double totalValue)
    {
        this.totalValue = totalValue;
    }

    // ===========================================================================

    @Override
    public String toString()
    {
        return ">> Day:\n"
        + "Date: " + date.format(FORMATTER) + "\n"
        + "Available Hours: " + availableHours + "\n"
        + "Optional Tasks: " + optionalTasks + "\n"
        + "Schedule: " + schedule + "\n"
        + "Total Value: " + totalValue + "\n"
        + "###################################\n";
    }
}

File: Server/src/services/day.ts
import { Period, printPeriod } from './period'
import { Constraint } from './constraint'
import { Task, printTask } from './task'

interface Day {
    date: Date
    optionalTasks: Task[]
    schedule: Task[] // Should be Task | Constraint
    availableHours: number
    totalValue: number
}

const addDays = (date: Date, days: number) => {
    const copy = new Date(Number(date))
    copy.setDate(date.getDate() + days)
    return copy
}

const generateCalendar = (numOfDays: number): Day[] => {
    let calendar: Day[] = new Array()
    const today: Day = {
        date: new Date(),
        optionalTasks: new Array() as Task[],
        schedule: new Array() as Task[],
        availableHours: 10,
        totalValue: 0,
    }

    calendar.push(today)

    for (let i = 1; i < numOfDays; i++) {
        calendar.push({
            date: addDays(new Date(), i),
            optionalTasks: new Array() as Task[],
            schedule: new Array() as Task[],
            availableHours: 10,
            totalValue: 0,
        } as Day)
    }

    return calendar
}

const printDay = (day: Day): void => {
    console.log('>> Day:')
    console.log('Available Hours: ' + day.availableHours)
    console.log('Optional Tasks: ' + day.optionalTasks.forEach((task) => printTask))
    console.log('Schedule: ' + day.schedule.forEach((period) => printTask))
    console.log('Total Value: ' + day.totalValue)
    console.log('###################################')
}

export { Day, generateCalendar, printDay }
